File: integration_tests/java_client_tests/client-app/src/test/java/dev/graphlink/test/UploadTest.java
package dev.graphlink.test;

import com.fasterxml.jackson.databind.ObjectMapper;
import dev.graphlink.test.generated.client.GLUpload;
import dev.graphlink.test.generated.client.GraphLinkClient;
import dev.graphlink.test.generated.types.UploadFileListResponse;
import dev.graphlink.test.generated.types.UploadOneFileResponse;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

import static dev.graphlink.test.Fixtures.*;
import static org.junit.jupiter.api.Assertions.*;

class UploadTest {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private MockAdapter adapter;
    private MockMultipartAdapter multipart;
    private GraphLinkClient client;

    private static GLUpload makeFile(int... ints) {
        byte[] bytes = new byte[ints.length];
        for (int i = 0; i < ints.length; i++) bytes[i] = (byte) ints[i];
        return GLUpload.fromBytes(bytes, "test.png", "image/png");
    }

    @BeforeEach
    void setUp() {
        adapter = new MockAdapter();
        multipart = new MockMultipartAdapter();
        client = newClientWithUpload(adapter, multipart, new MockWebSocketAdapter());
    }

    // ── uploadOneFile ─────────────────────────────────────────────────────────

    @Nested
    class UploadOneFile {

        @BeforeEach
        void register() {
            multipart.registerData("uploadOneFile", Map.of("uploadOneFile", kUserAliceJson));
        }

        @Test
        void response_id_isCorrect() throws Exception {
            UploadOneFileResponse res = client.mutations.uploadOneFile("user-1", makeFile(1, 2, 3));
            assertEquals("user-1", res.getUploadOneFile().getId());
        }

        @Test
        void response_name_isCorrect() throws Exception {
            UploadOneFileResponse res = client.mutations.uploadOneFile("user-1", makeFile(1, 2, 3));
            assertEquals("Alice Smith", res.getUploadOneFile().getName());
        }

        @Test
        void multipartAdapter_calledOnce() throws Exception {
            client.mutations.uploadOneFile("user-1", makeFile(1, 2, 3));
            assertEquals(1, multipart.getCallCount());
        }

        @Test
        void operations_containsCorrectUserId() throws Exception {
            client.mutations.uploadOneFile("user-42", makeFile(1, 2, 3));
            @SuppressWarnings("unchecked")
            Map<String, Object> ops = MAPPER.readValue(multipart.getLastCall().operations(), Map.class);
            @SuppressWarnings("unchecked")
            Map<String, Object> variables = (Map<String, Object>) ops.get("variables");
            assertEquals("user-42", variables.get("userId"));
        }

        @Test
        void fileVariable_isNull_inOperations() throws Exception {
            client.mutations.uploadOneFile("user-1", makeFile(1, 2, 3));
            @SuppressWarnings("unchecked")
            Map<String, Object> ops = MAPPER.readValue(multipart.getLastCall().operations(), Map.class);
            @SuppressWarnings("unchecked")
            Map<String, Object> variables = (Map<String, Object>) ops.get("variables");
            assertNull(variables.get("file"));
        }

        @Test
        void mapSlot0_pointsTo_variablesFile() throws Exception {
            client.mutations.uploadOneFile("user-1", makeFile(1, 2, 3));
            @SuppressWarnings("unchecked")
            Map<String, Object> map = MAPPER.readValue(multipart.getLastCall().mapJson(), Map.class);
            assertEquals(List.of("variables.file"), map.get("0"));
        }

        @Test
        void filePart_presentUnderSlot0() throws Exception {
            client.mutations.uploadOneFile("user-1", makeFile(1, 2, 3));
            assertTrue(multipart.getLastCall().files().containsKey("0"));
        }

        @Test
        void regularAdapter_notCalled() throws Exception {
            client.mutations.uploadOneFile("user-1", makeFile(1, 2, 3));
            assertEquals(0, adapter.getCallCount());
        }
    }

    // ── uploadFileList ────────────────────────────────────────────────────────

    @Nested
    class UploadFileList {

        @BeforeEach
        void register() {
            multipart.registerData("uploadFileList", Map.of("uploadFileList", kUserAliceJson));
        }

        @Test
        void response_id_isCorrect() throws Exception {
            UploadFileListResponse res = client.mutations.uploadFileList(
                    "user-1", List.of(makeFile(1, 2, 3), makeFile(4, 5, 6)));
            assertEquals("user-1", res.getUploadFileList().getId());
        }

        @Test
        void filesVariables_areAllNull_inOperations() throws Exception {
            client.mutations.uploadFileList(
                    "user-1", List.of(makeFile(1, 2, 3), makeFile(4, 5, 6)));
            @SuppressWarnings("unchecked")
            Map<String, Object> ops = MAPPER.readValue(multipart.getLastCall().operations(), Map.class);
            @SuppressWarnings("unchecked")
            Map<String, Object> variables = (Map<String, Object>) ops.get("variables");
            assertEquals(Arrays.asList(null, null), variables.get("files"));
        }

        @Test
        void mapSlot0_pointsTo_variablesFiles0() throws Exception {
            client.mutations.uploadFileList(
                    "user-1", List.of(makeFile(1, 2, 3), makeFile(4, 5, 6)));
            @SuppressWarnings("unchecked")
            Map<String, Object> map = MAPPER.readValue(multipart.getLastCall().mapJson(), Map.class);
            assertEquals(List.of("variables.files.0"), map.get("0"));
        }

        @Test
        void mapSlot1_pointsTo_variablesFiles1() throws Exception {
            client.mutations.uploadFileList(
                    "user-1", List.of(makeFile(1, 2, 3), makeFile(4, 5, 6)));
            @SuppressWarnings("unchecked")
            Map<String, Object> map = MAPPER.readValue(multipart.getLastCall().mapJson(), Map.class);
            assertEquals(List.of("variables.files.1"), map.get("1"));
        }

        @Test
        void bothFileParts_present() throws Exception {
            client.mutations.uploadFileList(
                    "user-1", List.of(makeFile(1, 2, 3), makeFile(4, 5, 6)));
            Map<String, ?> files = multipart.getLastCall().files();
            assertTrue(files.containsKey("0"));
            assertTrue(files.containsKey("1"));
        }

        @Test
        void singleFileList_producesOneSlot() throws Exception {
            client.mutations.uploadFileList("user-1", List.of(makeFile(1, 2, 3)));
            @SuppressWarnings("unchecked")
            Map<String, Object> map = MAPPER.readValue(multipart.getLastCall().mapJson(), Map.class);
            assertEquals(1, map.size());
        }
    }
}

File: integration_tests/kotlin_client_tests/client-app/src/test/kotlin/dev/graphlink/test/UploadTest.kt
package dev.graphlink.test

import dev.graphlink.test.generated.client.GLUpload
import dev.graphlink.test.generated.client.GraphLinkClient
import dev.graphlink.test.generated.client.KotlinxSerializationGraphLinkJsonCodec
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Nested
import kotlin.test.*
import dev.graphlink.test.Fixtures.kUserAliceJson
import dev.graphlink.test.Fixtures.newClientWithUpload

class UploadTest {

    private val codec = KotlinxSerializationGraphLinkJsonCodec()

    private lateinit var adapter: MockAdapter
    private lateinit var multipart: MockMultipartAdapter
    private lateinit var client: GraphLinkClient

    private fun makeFile(vararg bytes: Byte): GLUpload = GLUpload(
        stream = bytes.inputStream(),
        mimeType = "image/png",
        length = bytes.size.toLong(),
        filename = "test.png",
    )

    @BeforeTest
    fun setUp() {
        adapter = MockAdapter()
        multipart = MockMultipartAdapter()
        client = newClientWithUpload(adapter, MockWebSocketAdapter(), multipart)
    }

    // ── uploadOneFile ─────────────────────────────────────────────────────────

    @Nested
    inner class UploadOneFile {

        @BeforeTest
        fun register() {
            multipart.registerData("uploadOneFile", mapOf("uploadOneFile" to kUserAliceJson))
        }

        @Test
        fun response_id_isCorrect() = runTest {
            val res = client.mutations.uploadOneFile("user-1", makeFile(1, 2, 3))
            assertEquals("user-1", res.uploadOneFile.id)
        }

        @Test
        fun response_name_isCorrect() = runTest {
            val res = client.mutations.uploadOneFile("user-1", makeFile(1, 2, 3))
            assertEquals("Alice Smith", res.uploadOneFile.name)
        }

        @Test
        fun multipartAdapter_calledOnce() = runTest {
            client.mutations.uploadOneFile("user-1", makeFile(1, 2, 3))
            assertEquals(1, multipart.getCallCount())
        }

        @Test
        fun operations_containsCorrectUserId() = runTest {
            client.mutations.uploadOneFile("user-42", makeFile(1, 2, 3))
            val ops = codec.decode(multipart.getLastCall()!!.operations)
            @Suppress("UNCHECKED_CAST")
            val variables = ops["variables"] as Map<String, Any?>
            assertEquals("user-42", variables["userId"])
        }

        @Test
        fun fileVariable_isNull_inOperations() = runTest {
            client.mutations.uploadOneFile("user-1", makeFile(1, 2, 3))
            val ops = codec.decode(multipart.getLastCall()!!.operations)
            @Suppress("UNCHECKED_CAST")
            val variables = ops["variables"] as Map<String, Any?>
            assertNull(variables["file"])
        }

        @Test
        fun mapSlot0_pointsTo_variablesFile() = runTest {
            client.mutations.uploadOneFile("user-1", makeFile(1, 2, 3))
            val map = codec.decode(multipart.getLastCall()!!.mapJson)
            assertEquals(listOf("variables.file"), map["0"])
        }

        @Test
        fun filePart_presentUnderSlot0() = runTest {
            client.mutations.uploadOneFile("user-1", makeFile(1, 2, 3))
            assertTrue(multipart.getLastCall()!!.files.containsKey("0"))
        }

        @Test
        fun regularAdapter_notCalled() = runTest {
            client.mutations.uploadOneFile("user-1", makeFile(1, 2, 3))
            assertEquals(0, adapter.getCallCount())
        }
    }

    // ── uploadFileList ────────────────────────────────────────────────────────

    @Nested
    inner class UploadFileList {

        @BeforeTest
        fun register() {
            multipart.registerData("uploadFileList", mapOf("uploadFileList" to kUserAliceJson))
        }

        @Test
        fun response_id_isCorrect() = runTest {
            val res = client.mutations.uploadFileList("user-1", listOf(makeFile(1, 2, 3), makeFile(4, 5, 6)))
            assertEquals("user-1", res.uploadFileList.id)
        }

        @Test
        fun filesVariables_areAllNull_inOperations() = runTest {
            client.mutations.uploadFileList("user-1", listOf(makeFile(1, 2, 3), makeFile(4, 5, 6)))
            val ops = codec.decode(multipart.getLastCall()!!.operations)
            @Suppress("UNCHECKED_CAST")
            val variables = ops["variables"] as Map<String, Any?>
            val files = variables["files"] as List<*>
            assertEquals(2, files.size)
            assertNull(files[0])
            assertNull(files[1])
        }

        @Test
        fun mapSlot0_pointsTo_variablesFiles0() = runTest {
            client.mutations.uploadFileList("user-1", listOf(makeFile(1, 2, 3), makeFile(4, 5, 6)))
            val map = codec.decode(multipart.getLastCall()!!.mapJson)
            assertEquals(listOf("variables.files.0"), map["0"])
        }

        @Test
        fun mapSlot1_pointsTo_variablesFiles1() = runTest {
            client.mutations.uploadFileList("user-1", listOf(makeFile(1, 2, 3), makeFile(4, 5, 6)))
            val map = codec.decode(multipart.getLastCall()!!.mapJson)
            assertEquals(listOf("variables.files.1"), map["1"])
        }

        @Test
        fun bothFileParts_present() = runTest {
            client.mutations.uploadFileList("user-1", listOf(makeFile(1, 2, 3), makeFile(4, 5, 6)))
            val files = multipart.getLastCall()!!.files
            assertTrue(files.containsKey("0"))
            assertTrue(files.containsKey("1"))
        }

        @Test
        fun singleFileList_producesOneSlot() = runTest {
            client.mutations.uploadFileList("user-1", listOf(makeFile(1, 2, 3)))
            val map = codec.decode(multipart.getLastCall()!!.mapJson)
            assertEquals(1, map.size)
        }
    }
}
